lib/Runtime/Library/JavascriptPromise.h

```diff
@@ -538,28 +542,52 @@ namespace Js
         Var Resolve(Var resolution, ScriptContext* scriptContext);
         Var Reject(Var resolution, ScriptContext* scriptContext);
 
-        enum PromiseStatus
+        enum PromiseStatus : unsigned char
         {
-            PromiseStatusCode_Undefined,
-            PromiseStatusCode_Unresolved,
-            PromiseStatusCode_HasResolution,
-            PromiseStatusCode_HasRejection
+            PromiseStatusCode_Undefined     = 0x00,
+            PromiseStatusCode_Unresolved    = 0x01,
+            PromiseStatusCode_HasResolution = 0x02,
+            PromiseStatusCode_HasRejection  = 0x03
         };
 
-        bool GetIsHandled() { return isHandled; }
-        void SetIsHandled() { isHandled = true; }
-        PromiseStatus GetStatus() const { return status; }
+        static const unsigned char statusMask = 0x03;
+        static const unsigned char isHandledMask = 0x04;
+
+        bool GetIsHandled() const { return (this->state & isHandledMask) == isHandledMask; }
+        void SetIsHandled() 
+        {
+#ifdef DEBUG
+            PromiseStatus oldStatus = GetStatus();
+#endif
+            this->state |= isHandledMask;  
+#ifdef DEBUG
+            AssertMsg(this->GetStatus() == oldStatus, "Unexpected status after setIsHandled");
+            AssertMsg(this->GetIsHandled() == true, "Unexpected value after SetIsHandled"); 
+#endif
+        }
+        PromiseStatus GetStatus() const { return static_cast<PromiseStatus>(this->state & statusMask); }
+
+        void SetStatus(const PromiseStatus newStatus) {
+#ifdef DEBUG
+            bool oldIsHandled = GetIsHandled();
+#endif
+            this->state = (this->state & ~statusMask) | newStatus;
+#ifdef DEBUG
+            AssertMsg(this->GetStatus() == newStatus, "Unexpected status after setStatus");
+            AssertMsg(this->GetIsHandled() == oldIsHandled, "unexpected isHandled after setSTatus");
+#endif
+        }
+
         Var GetResult() const { return result; }
 
     protected:
         Var ResolveHelper(Var resolution, bool isRejecting, ScriptContext* scriptContext);
 
     protected:
-        Field(PromiseStatus) status;
-        Field(bool) isHandled;
+        //Field(PromiseStatus) status;
         Field(Var) result;
-        Field(JavascriptPromiseReactionList*) resolveReactions;
-        Field(JavascriptPromiseReactionList*) rejectReactions;
+        Field(JavascriptPromiseReactionList*) reactions;
+        Field(unsigned char) state;
```

Comment:
It turns out that packing the `isHandled` state and the `PromiseStatus` into a single byte doesn't buy us much since the compiler is aligning this on 8-byte boundaries in the 64-bit build, giving us a 56-byte size.  Also, the recycler puts this into the 64-byte bucket, so we have room here.  

I'm happy to have a separate `PromiseStatus` & `isHandled` as it will make this code a tad bit more clear.